src/testdrive/src/action.rs

```diff
@@ -446,12 +470,12 @@ pub fn build(cmds: Vec<PosCommand>, state: &State) -> Result<Vec<PosAction>, Err
                         }
                     }
                 }
-                Box::new(sql::build_sql(sql, sql_timeout).map_err(wrap_err)?)
+                Box::new(sql::build_sql(sql, sql_context.clone()).map_err(wrap_err)?)
```

Comment:
To see why you are forced to clone `sql_context` here it helps to start from the return value of the function. The functions returns a `Vec<PosAction>` which will contain several independent `PosAction` objects. Each of these contain the `Action` returned by `build_sql`, and each `Action` wants to use some kind of `SqlContext`.

Passing just `sql_context` would work fine if it wasn't for the for loop. But because you can only move a value once you cannot have iteration because the second iteration of the loop would have nothing to move.

Passing a `&sql_context` doesn't work either because it is declared on the stack. If that was allowed the items in the Vec<PosAction> would contain a dangling pointer. 

Allocating `sql_context` on the heap and then passing `&sql_context` wouldn't work either! That is because even though you might be careful to not deallocate the heap allocation while the elements of the Vec are still around, rustc cannot prove that so it's not allowed.

So the only two options are to either clone the whole thing, like you did, or you can opt for a reference counted wrapper type (`Rc<T>` or `Arc<T>`) that turn clones into increments and decrements of an internal reference counter and automatically drop the inner T once the last one of the clones is dropped.

I hope this helped!

Reply:
Petros brings the explanation of what your options are: I would like to ask one more thing. (:

Do we need to have one sql context across all the actions being constructed here, or is it reasonable for each action to have its own sql context?

If it's the former, then yup, you shouldn't clone & you'll need a heap-allocated context to modify that as if it were a singleton. If the latter, then I think your .clone is correct & necessary!

Reply:
@petrosagg thank you for the detailed explanation! I need to reconsider my life choices ...

Reply:
@antifuchs ```Box``` does not work , as once the struct is boxed individual members can not be modified, and I need to be able to modify the regexp as new ```set-timeout``` directives are processed. So I will stick with ```clone()```

Reply:
@antifuchs I am afraid I could not get it to work -- rustc asks me to add explicit static lifetimes. All of this is beyond my rust pay grade right now , so if you all do not mind I will stick with ```clone()```

Reply:
I've got a diff for Arc-ifying the context here if you want to take a look (it's a bit long but the result compiles):

<details>
<summary>a diff for wrapping sql_context in an Arc</summary>

```diff
diff --git a/src/testdrive/src/action.rs b/src/testdrive/src/action.rs
index c5606ab72..d384eca4d 100644
--- a/src/testdrive/src/action.rs
+++ b/src/testdrive/src/action.rs
@@ -7,12 +7,15 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-use std::collections::{BTreeSet, HashMap};
 use std::fs;
 use std::future::Future;
 use std::net::ToSocketAddrs;
 use std::path::PathBuf;
 use std::time::Duration;
+use std::{
+    collections::{BTreeSet, HashMap},
+    sync::Arc,
+};
 
 use async_trait::async_trait;
 use futures::future::FutureExt;
@@ -297,11 +300,11 @@ where
 pub fn build(cmds: Vec<PosCommand>, state: &State) -> Result<Vec<PosAction>, Error> {
     let mut out = Vec::new();
     let mut vars = HashMap::new();
-    let mut sql_context = SQLContext {
+    let mut sql_context = Arc::new(SQLContext {
         sql_timeout: DEFAULT_SQL_TIMEOUT,
         regex: None,
         regex_replacement: DEFAULT_REGEX_REPLACEMENT.to_string(),
-    };
+    });
 
     vars.insert("testdrive.kafka-addr".into(), state.kafka_addr.clone());
     vars.insert(
@@ -421,22 +424,23 @@ pub fn build(cmds: Vec<PosCommand>, state: &State) -> Result<Vec<PosAction>, Err
                     }
                     "set-regex" => {
                         let regex_str = builtin.args.string("match").map_err(wrap_err)?;
-                        sql_context.regex = Some(Regex::new(&regex_str).unwrap());
+                        (*sql_context).regex = Some(Regex::new(&regex_str).unwrap());
 
                         let regex_replacement_opt = builtin.args.opt_string("replacement");
                         if let Some(regex_replacement) = &regex_replacement_opt {
-                            sql_context.regex_replacement = regex_replacement.to_string();
+                            (*sql_context).regex_replacement = regex_replacement.to_string();
                         } else {
-                            sql_context.regex_replacement = DEFAULT_REGEX_REPLACEMENT.to_string();
+                            (*sql_context).regex_replacement =
+                                DEFAULT_REGEX_REPLACEMENT.to_string();
                         }
                         continue;
                     }
                     "set-sql-timeout" => {
                         let duration = builtin.args.string("duration").map_err(wrap_err)?;
                         if duration.to_lowercase() == "default" {
-                            sql_context.sql_timeout = DEFAULT_SQL_TIMEOUT;
+                            (*sql_context).sql_timeout = DEFAULT_SQL_TIMEOUT;
                         } else {
-                            sql_context.sql_timeout = parse_duration::parse(&duration)
+                            (*sql_context).sql_timeout = parse_duration::parse(&duration)
                                 .map_err(|e| wrap_err(e.to_string()))?;
                         }
                         continue;
@@ -470,12 +474,12 @@ pub fn build(cmds: Vec<PosCommand>, state: &State) -> Result<Vec<PosAction>, Err
                         }
                     }
                 }
-                Box::new(sql::build_sql(sql, sql_context.clone()).map_err(wrap_err)?)
+                Box::new(sql::build_sql(sql, Arc::clone(sql_context)).map_err(wrap_err)?)
             }
             Command::FailSql(mut sql) => {
                 sql.query = subst(&sql.query)?;
                 sql.expected_error = subst(&sql.expected_error)?;
-                Box::new(sql::build_fail_sql(sql, sql_context.clone()).map_err(wrap_err)?)
+                Box::new(sql::build_fail_sql(sql, Arc::clone(sql_context)).map_err(wrap_err)?)
             }
         };
         out.push(PosAction {
diff --git a/src/testdrive/src/action/sql.rs b/src/testdrive/src/action/sql.rs
index 5388884d6..1e87656d1 100644
--- a/src/testdrive/src/action/sql.rs
+++ b/src/testdrive/src/action/sql.rs
@@ -7,10 +7,10 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-use std::ascii;
 use std::error::Error;
 use std::fmt::{self, Write as _};
 use std::io::{self, Write};
+use std::{ascii, sync::Arc};
 
 use async_trait::async_trait;
 use md5::{Digest, Md5};
@@ -34,10 +34,10 @@ use crate::parser::{FailSqlCommand, SqlCommand, SqlOutput};
 pub struct SqlAction {
     cmd: SqlCommand,
     stmt: Statement<Raw>,
-    sql_context: SQLContext,
+    sql_context: Arc<SQLContext>,
 }
 
-pub fn build_sql(mut cmd: SqlCommand, sql_context: SQLContext) -> Result<SqlAction, String> {
+pub fn build_sql(mut cmd: SqlCommand, sql_context: Arc<SQLContext>) -> Result<SqlAction, String> {
     let stmts = sql_parser::parser::parse_statements(&cmd.query)
         .map_err(|e| format!("unable to parse SQL: {}: {}", cmd.query, e))?;
     if stmts.len() != 1 {
@@ -270,17 +270,14 @@ impl SqlAction {
 
 pub struct FailSqlAction {
     cmd: FailSqlCommand,
-    sql_context: SQLContext,
+    sql_context: Arc<SQLContext>,
 }
 
 pub fn build_fail_sql(
     cmd: FailSqlCommand,
-    sql_context: SQLContext,
+    sql_context: Arc<SQLContext>,
 ) -> Result<FailSqlAction, String> {
-    Ok(FailSqlAction {
-        cmd,
-        sql_context: sql_context,
-    })
+    Ok(FailSqlAction { cmd, sql_context })
 }
 
 #[async_trait]
```
</details>

All that said/pasted - the thing we need to think about is whether the .clone solution is what's needed, or the Arc/Rc one: With the Arc solution, changes to the sql context are "global" (that is, any changes to the context would apply to statements before the changes are made); with the .clone, they affect only the statements that come after.

Reply:
@antifuchs right now all ```$``` directives in the ```.td``` file are processed at startup, regardless of their position in the file. However, I am hoping that eventually those directives will be processed at their original place within the test. So, in the longer term, ```.clone()``` is what seems to be the more natural choice.